File: heads/haskell/src/main/haskell/Hydra/Lib/Strings.hs
-- | Haskell implementations of hydra.lib.strings primitives

module Hydra.Lib.Strings where

import qualified Data.Char as C
import qualified Data.List as L
import qualified Data.List.Split as LS


-- | Concatenate a list of strings into a single string.
cat :: [String] -> String
cat = L.concat

-- | Concatenate two strings.
cat2 :: String -> String -> String
cat2 s1 s2 = s1 ++ s2

-- Bootstrap stub: removed in Phase 2.4 but kept so stale dist modules
-- continue to compile during Step 1. Not registered as a primitive.
charAt :: Int -> String -> Int
charAt i s = C.ord (s !! i)

-- | Convert a list of Unicode code points to a string.
fromList :: [Int] -> String
fromList = fmap C.chr

-- | Join a list of strings with a separator between each element.
intercalate :: String -> [String] -> String
intercalate = L.intercalate

-- | Return the length of a string.
length :: String -> Int
length = L.length

-- | Get the Unicode code point of the character at a specific index, returning Nothing if out of bounds.
maybeCharAt :: Int -> String -> Maybe Int
maybeCharAt i s
  | i < 0 || i >= L.length s = Nothing
  | otherwise = Just (C.ord (s !! i))

-- | Split a string into lines.
lines :: String -> [String]
lines = L.lines

-- | Check whether a string is empty.
null :: String -> Bool
null = L.null

-- | Split a string on a delimiter string.
splitOn :: String -> String -> [String]
splitOn = LS.splitOn

-- | Convert a string to a list of Unicode code points.
toList :: String -> [Int]
toList = fmap C.ord

-- | Convert a string to lowercase.
toLower :: String -> String
toLower = fmap C.toLower

-- | Convert a string to uppercase.
toUpper :: String -> String
toUpper = fmap C.toUpper

-- | Join a list of strings with newlines, appending a trailing newline.
unlines :: [String] -> String
unlines = L.unlines

File: heads/typescript/src/main/typescript/hydra/lib/strings.ts
// Hand-written runtime: hydra.lib.strings primitives.
//
// Operates on JavaScript strings. Code points are represented as `number`,
// matching Hydra's `Int` for characters.

export const cat = (xs: readonly string[]): string => xs.join("");
export const cat2 = (a: string) => (b: string): string => a + b;

export const intercalate = (sep: string) => (xs: readonly string[]): string => xs.join(sep);

// Counts Unicode code points (not UTF-16 code units), matching the
// kernel test fixtures' "length of string" expectation.
export const length = (s: string): number => [...s].length;

// `splitOn` mirrors Haskell's Data.List.Split.splitOn. The empty
// separator case is special: it yields `["", c0, c1, ..., cN-1, ""]` —
// the input split between every code point with empty boundary strings.
// For non-empty seps we delegate to JS's String.split.
export const splitOn = (sep: string) => (s: string): readonly string[] => {
  if (sep === "") {
    if (s === "") return [""];
    return ["", ...s.split("")];
  }
  return s.split(sep);
};

export const fromList = (cps: readonly number[]): string => String.fromCodePoint(...cps);

export const toList = (s: string): readonly number[] => {
  const out: number[] = [];
  for (const ch of s) out.push(ch.codePointAt(0)!);
  return out;
};

export const null_ = (s: string): boolean => s.length === 0;

export const toUpper = (s: string): string => s.toUpperCase();
export const toLower = (s: string): string => s.toLowerCase();

export const trim = (s: string): string => s.trim();

// `lines` mirrors Haskell's `Data.List.lines`: an empty input yields
// `[]`; a trailing newline does NOT add an empty final element.
export const lines = (s: string): readonly string[] => {
  if (s === "") return [];
  const parts = s.split("\n");
  // Drop trailing empty caused by a final "\n".
  if (parts[parts.length - 1] === "") parts.pop();
  return parts;
};
// `unlines` appends "\n" after EVERY element (including the last),
// matching Haskell's `Data.List.unlines`.
export const unlines = (xs: readonly string[]): string => xs.map((x) => x + "\n").join("");

export const maybeCharAt = (i: number) => (s: string): import("../core.js").Maybe<number> => {
  const cp = s.codePointAt(i);
  return cp === undefined ? { tag: "nothing" } : { tag: "just", value: cp };
};

export const isPrefix = (pre: string) => (s: string): boolean => s.startsWith(pre);
export const isSuffix = (suf: string) => (s: string): boolean => s.endsWith(suf);
